samza-core/src/main/scala/org/apache/samza/config/StreamConfig.java

```diff
@@ -0,0 +1,394 @@
+    if (!property.startsWith(SAMZA_PROPERTY)) {
+      throw new IllegalArgumentException(
+        String.format("Attempt to fetch a non samza property for SystemStream %s named %s", systemStream, property));
+    }
+
+    String streamVal = getAllStreamProperties(systemStreamToStreamId(systemStream)).get(property);
+    return (streamVal != null) ? streamVal : getSystemStreamProperties(systemStream.getSystem(), systemStream.getStream()).get(property);
+  }
+
+  /**
+   * Gets the System associated with the specified streamId.
+   * It first looks for the property
+   * streams.{streamId}.system
+   * <p>
+   * If no value was provided, it uses
+   * job.default.system
+   *
+   * @param streamId the identifier for the stream in the config.
+   * @return the system name associated with the stream.
+   */
+  public String getSystem(String streamId) {
+    return getOrDefault(String.format(SYSTEM_FOR_STREAM_ID, streamId),
+      new JobConfig(this).getDefaultSystem().orElse(null));
+  }
+
+  /**
+   * Gets the physical name for the specified streamId.
+   *
+   * @param streamId the identifier for the stream in the config.
+   * @return the physical identifier for the stream or the default if it is undefined.
+   */
+  public String getPhysicalName(String streamId) {
+    // use streamId as the default physical name
+    return getOrDefault(String.format(PHYSICAL_NAME_FOR_STREAM_ID, streamId), streamId);
+  }
+
+
+  public Optional<String> getStreamMsgSerde(SystemStream systemStream) {
+    return nonEmptyOption(getSamzaProperty(systemStream, MSG_SERDE));
+  }
+
+  public Optional<String> getStreamKeySerde(SystemStream systemStream) {
+    return nonEmptyOption(getSamzaProperty(systemStream, KEY_SERDE));
+  }
+
+  public boolean getResetOffset(SystemStream systemStream) {
+    String resetOffset = getSamzaProperty(systemStream, CONSUMER_RESET_OFFSET, "false");
+    if (!resetOffset.equalsIgnoreCase("true") && !resetOffset.equalsIgnoreCase("false")) {
+      LOG.warn("Got a .samza.reset.offset configuration for SystemStream {} that is not true or false (was {})." +
+        " Defaulting to false.", systemStream, resetOffset);
+
+      resetOffset = "false";
+    }
+    return Boolean.valueOf(resetOffset);
+  }
+
+
+  /**
+   * Gets a Samza property, with a default value used if no property value is found.
+   * See getSamzaProperty(SystemStream, String).
+   *
+   * @param systemStream the SystemStream for which the property value will be retrieved.
+   * @param property the samza property key (including the "samza." prefix); for example, for both
+   *                 "streams.streamId.samza.prop.key" and "systems.system.streams.streamName.samza.prop.key", this
+   *                 argument should have the value "samza.prop.key"
+   * @param defaultValue the default value to use if the property value is not found
+   */
+  protected String getSamzaProperty(SystemStream systemStream, String property, String defaultValue) {
```

Comment:
Can `getSamzaProperty` (here and below) be private?
Could you please move these helpers further below in the class file so that they are not in between some of the public methods?